config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java

```diff
@@ -112,15 +112,24 @@
                 String configPropName = originalNames.get(beanProp.getName());
                 // Is the property key missing in the config?
                 if (configPropName == null) {
-                    // If so, continue if the field is marked as @{link Optional}
-                    if (isOptionalProperty(clazz, beanProp)) {
-                        continue;
+                    // If the field is not marked as @{link Optional}...
+                    if (!isOptionalProperty(clazz, beanProp)) {
+                        // raise a {@link Missing} exception right here
+                        throw new ConfigException.Missing(beanProp.getName());
+                    }
+                } else {
+                    try {
+                        Object unwrapped = getValue(
```

Comment:
I wonder if it would be better to handle null in getValue (would need to pass in to getValue whether the prop is optional and then use `config.getIsNull` I think). An advantage would be to avoid exceptions in normal control flow, which is nice because they are legitimately pretty slow to throw, and that could matter to some programs.